packages/backend-app-api/src/wiring/BackendInitializer.ts

```diff
@@ -147,24 +156,84 @@ export class BackendInitializer {
           }
         }
 
-        this.#registerInits.push({
-          id: r.type === 'plugin' ? r.pluginId : `${r.pluginId}.${r.moduleId}`,
-          provides,
-          consumes: new Set(Object.values(r.init.deps)),
-          init: r.init,
-        });
+        if (r.type === 'plugin') {
+          if (pluginInits.has(r.pluginId)) {
+            throw new Error(`Plugin '${r.pluginId}' is already registered`);
+          }
+          pluginInits.set(r.pluginId, {
+            provides,
+            consumes: new Set(Object.values(r.init.deps)),
+            init: r.init,
+          });
+        } else {
+          let modules = moduleInits.get(r.pluginId);
+          if (!modules) {
+            modules = new Map();
+            moduleInits.set(r.pluginId, modules);
+          }
+          if (modules.has(r.moduleId)) {
+            throw new Error(
+              `Module '${r.moduleId}' for plugin '${r.pluginId}' is already registered`,
+            );
+          }
+          modules.set(r.moduleId, {
+            provides,
+            consumes: new Set(Object.values(r.init.deps)),
+            init: r.init,
+          });
+        }
       }
     }
 
-    const orderedRegisterResults = this.#resolveInitOrder(this.#registerInits);
+    const allPluginIds = [
+      ...new Set([...pluginInits.keys(), ...moduleInits.keys()]),
+    ];
 
-    for (const registerInit of orderedRegisterResults) {
-      const deps = await this.#getInitDeps(
-        registerInit.init.deps,
-        registerInit.id,
-      );
-      await registerInit.init.func(deps);
-    }
+    // All plugins are initialized in parallel
+    await Promise.all(
+      allPluginIds.map(async pluginId => {
+        // Modules are initialized before plugins, so that they can provide extension to the plugin
+        const modules = moduleInits.get(pluginId) ?? [];
+        await Promise.all(
+          Array.from(modules).map(async ([moduleId, moduleInit]) => {
+            const moduleDeps = await this.#getInitDeps(
+              moduleInit.init.deps,
+              pluginId,
+            );
+            await moduleInit.init.func(moduleDeps).catch(error => {
+              throw new ForwardedError(
+                `Module '${moduleId}' for plugin '${pluginId}' startup failed`,
+                error,
+              );
+            });
+          }),
+        );
+
+        // Once all modules have been initialized, we can initialize the plugin itself
```

Comment:
I think that maybe it's fine, but I'm reading that if a module uses the same `db`, theres a bit of a race condition that the migrations won't have completed yet?

Reply:
Oh yeah the main plugin migrations will not have completed by the time each module is initialized. This is a bit of a constraint in the design of the system I'd say and also existed before this change. I think the new startup hook is likely a good way to handle startup code that needs to run after the plugin is initialized, but haven't thought about it extensively

Reply:
Good observation.

For e.g. entity providers that's handled through having the separate `connect` happening later, but I bet this will bite someone somewhere 🤔